### ISSUE TITLE
`PhysicsDirectSpaceState3D`'s `intersect_ray`, `intersect_point`, and `intersect_shape` all ignore `GridMap`s

### ISSUE THREAD

**Issue by jtnicholl** [godotengine/godot#76349]
### Godot version

v4.0.2.stable.official [7a0977ce2]

### System information

Fedora 37

### Issue description

The methods `intersect_ray`, `intersect_point`, and `intersect_shape` from `PhysicsDirectSpaceState3D` do not take `GridMap`s from the scene into consideration, and will not collide.

### Steps to reproduce

1. Try to check for a collision with a GridMap using one of the aforementioned methods.
2. No collision is ever reported.

### Minimal reproduction project

[GridmapQueries.zip](https://github.com/godotengine/godot/files/11301948/GridmapQueries.zip)

**Comment by smix8** [comment]
You can not query the physics space in a _ready() function at game start. The PhysicsServer had not time to sync, same with Navigation or all kinds of other servers. The _ready() function of nodes is not the right place to do server queries at game start.

**Comment by jtnicholl** [comment]
> You can not query the physics space in a _ready() function at game start. The PhysicsServer had not time to sync, same with Navigation or all kinds of other servers. The _ready() function of nodes is not the right place to do server queries at game start.

That is odd, because it does work in the ready function with StaticBody nodes, just not GridMaps. I have updated the MRP to demonstrate this.

This fact should probably be documented better, none of the method or property descriptions for `direct_space_state` or the methods mention it.

**Comment by dashdotdashdot** [comment]
> If you move your query to a deferred call or a process function it works just fine.

This was not the case for me - see my duplicate #77872 - when I raycast from inside my scene's physics processing function. During the first physics processing callback (frame 1?), my GridMap did not collide with my raycast; it only collided from the second physics processing callback onward.


> That is odd, because it does work in the ready function with StaticBody nodes, just not GridMaps.

That other PhysicsBody3D-derived nodes *can* be hit by raycasts during the first physics processing callback after entering the tree is what lead me to believe that my GridMaps not being hit was erroneous.


> This fact should probably be documented better, none of the method or property descriptions for `direct_space_state` or the methods mention it.

I agree, having just spent considerable time in trying to troubleshoot it, thinking I had done something incorrectly, and then in building an MRP to report it as an issue.

It seems reasonable for a user to believe that collision should work from the first physics processing callback after a scene has become ready - after all, is that not what "ready" suggests? That collision does not work this way seems to violate the principle of least astonishment, making it especially important to document exceptions like GridMaps and whatever else might also be affected.

**Comment by dashdotdashdot** [comment]
> I'd recommend in these cases to search for bug reports

That was, in fact, close to the first thing I did. Alas, I searched for `is:issue is:open gridmap raycast` and the like - you'll notice that the first instance of the latter keyword in this issue is my first comment, so I wasn't able to find it until now. Nor did I find anything using search engines except several unanswered questions on https://godotengine.org/qa/questions and the subreddit.

I searched the best I could with the symptoms I observed, and now that I've been directed here I'm not complaining, simply commenting to give this hypothetical "someone" further context on the open documentation issue, that being that a reasonable person acting in good faith can and has fallen into this snare. No need to snip.

**Comment by AThousandShips** [comment]
I'm not snipping, I was making a suggestion to spare yourself time, I didn't assume any intent from you, so please do me the same courtesy, always assume good intent

**Comment by smix8** [comment]
The issue with relying on "ready" as a notification is that it is just a notification that gets send after all the other notifactions like "enter tree" or "enter world"  in the loop and it does not check anything.

If the node has a more complex or threaded first time setup it is not blocking the notification loop so "ready" is send even while technically the node has not finished its entire full setup.

**Comment by dashdotdashdot** [comment]
> The issue with relying on "ready" as a notification is that ... it does not check anything.

> "ready" is send even while technically the node has not finished its entire full setup.

At risk of duplicating my question in #77872, is there a way to know when a given node has become truly ready, i.e. has finished its entire full setup?

**Comment by dashdotdashdot** [comment]
> > Is it that process_frame's first emission is related to whether a scene's nodes (e.g. GridMap) are "truly ready", or simply that awaiting it gives us a good chance of having waited long enough?
> 
> `await get_tree().process_frame` just means "wait 1 frame before running the code below this line". It's a way to delay code without blocking execution.

Okay, so it's simply a way to delay for a frame. That's important to know, and I'm grateful for the clarification. So does this then mean that there's presently no way to know when a GridMap has been fully set up and may be collided with? Or does it mean that waiting a frame is sufficient to guarantee that the GridMap has been fully set up and may be collided with?

**Comment by dashdotdashdot** [comment]
> You keep asking the same question that has already been answered:

We point out that GridMap cannot be collided with when one might reasonably expect it to be - why? Well, we're told, GridMap et al.'s setup is deferred and there's no "standardized" way "depending on the node" to know when it is, so we're to await the SceneTree's `process_frame` signal for "one or more frames". The solution given in the latter, in the context of the ambiguity introduced by the former, is necessary but not sufficient.

I would not be asking for clarification if the question had already been answered: Is it possible, *specifically* to know when GridMap can be collided with? If the answer is "no", then that's fine because it can be worked around, but "await process_frame" does not close it one way or the other.

**Comment by AThousandShips** [comment]
To clarify, you are saying that the situation that the OP is seeing, that it does work when waiting one frame, is not the case for you? That the suggestion given does not work reliably for you? Because otherwise that's the simple answer, also are you only awaiting `process_frame`? Because as said over in the other issue physics frames might be skipped etc.

**Comment by dashdotdashdot** [comment]
> To clarify, you are saying that the situation that the OP is seeing, that it does work when waiting one frame, is not the case for you? That the suggestion given does not work reliably for you? Because otherwise that's the simple answer, also are you only awaiting `process_frame`? Because as said over in the other issue physics frames might be skipped etc.

We are told that the setup is deferred and that (reading expansively) it may take "one or more frames" to complete, which makes sense given the constraints explained above. In my own, most simple case, I see GridMap become collidable on the second physics frame. Which is wonderful news for these tiny little tests, but will it *always* take one physics frame, no matter the size and circumstance of the GridMap in question?

Common sense, knowledge, and experience says "probably not", so the question then becomes "if the time before it's ready is unknown, can we know when it's finished?". *That* question was answered with "depends; defer by awaiting process_frame". To which the obvious clarification is "okay, but what will that do for us [in the context of needing to know when the GridMap is ready to be collided with]?" We are presently here.

Again, if the answer is "nothing; it's presently completely unknown and unknowable, you just have to wait until you're pretty sure it's done and then confirm by other means" then that's at least manageable.

**Comment by AThousandShips** [comment]
Can you confirm that it *doesn't* become ready on the second frame? Because everything points to that it will be, and there's no indication it doesn't, and making the assumption it won't doesn't seem based on anything, the code doesn't indicate that anything is delayed further, from what I can see the delay is from the server setting things up on the first frame, I can't find any place where `GridMap` itself defers anything, so it looks like a safe assumption.

I understand the need to be certain here, but asking the same question multiple times and getting responses on the line of "it depends" or "it should be this way" isn't productive and isn't going to give you definitive answers, you know?

To be clear:
There's no reason to delay visuals or most things until everything is ready, doing so would cause delays that people don't want, some components and features defer doing things for this very reason

So, in conclusion:
Do some testing yourself, if you can't get a clear answer in the short time frame you've been asking here, and if there is indeed a variation to the required delay, then that indicates that `GridMap` needs its own signal for ready. Now again, from what I can see it doesn't seem that this is the case, and I've seen no reason here to believe it is not, not from the code or the reported experiences. Again, caution is good, but I don't see there's any reason to assume this isn't how it works, and asking for complete reassurance in this way isn't productive,

**Comment by Theyoungster** [comment]
Godot 4.3.stable, lost 4 hours before I figured out this was happening. _Aaaaaaaaaaaaaaaaaaaa._

> await get_tree().process_frame

This worked at least.
